### PR TITLE
zing: init at 24.03.0.0

### PR DESCRIPTION
**Pull request by TerrorJack**
## Description of changes

zing is a famous proprietary jdk implementation by azul that features low pause times. I use it to run some java stuff on my own vps and might be useful to other folks as well.

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [x] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by mweinelt** [comment]
Triggering eval after it was fixed on master.

@ofborg eval

**Comment by TerrorJack** [comment]
Result of `nixpkgs-review pr 256354` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>zing</li>
  </ul>
</details>

**Comment by TerrorJack** [comment]
Result of `nixpkgs-review pr 256354` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>zing</li>
  </ul>
</details>

**Comment by TerrorJack** [comment]
Result of `nixpkgs-review pr 256354` run on aarch64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>zing</li>
  </ul>
</details>

**Comment by TerrorJack** [comment]
Result of `nixpkgs-review pr 256354` run on aarch64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>zing</li>
  </ul>
</details>

**Comment by TerrorJack** [comment]
Result of `nixpkgs-review pr 256354` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>zing</li>
  </ul>
</details>

File: pkgs/development/compilers/zing/default.nix [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+      lib.escapeShellArg ''
+        public class HelloWorld {
+          public static void main(String[] args) {
+            System.out.println("Hello World");
+          }
+        }
+      ''
+    } > HelloWorld.java
+
+    $out/bin/javac HelloWorld.java
```
Comment (h7x4) [inline pkgs/development/compilers/zing/default.nix:67]: I think this looks a bit cleaner, but it's not very important.

```suggestion
  helloWorldJava = ''
    public class HelloWorld {
      public static void main(String[] args) {
        System.out.println("Hello World");
      }
    }
  '';
  passAsFile = [ "helloWorldJava" ];

  doInstallCheck = stdenvNoCC.buildPlatform.system
    == stdenvNoCC.hostPlatform.system;
  installCheckPhase = ''
    runHook preInstallCheck

    pushd "$(mktemp -d)"

    cp "$helloWorldJavaPath" HelloWorld.java
    $out/bin/javac HelloWorld.java
```

File: pkgs/development/compilers/zing/default.nix [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+, fetchzip
+, lib
+, setJavaClassPath
+, stdenv
+, stdenvNoCC
+, xorg
+, zlib
+,
+}:
+stdenvNoCC.mkDerivation (finalAttrs: {
```
[the remark this answers is not in the document]
Comment (TerrorJack) [reply pkgs/development/compilers/zing/default.nix]: yup, thanks for pointing out :)

**Comment by FliegendeWurst** [comment]
@NixOS/java: are we interested in having zing?

**Comment by tomodachi94** [comment]
I'm hesitant to include a new JDK, but this one looks like it has tangible benefits not present in some distributions. I'm conflicted.

**Comment by emilazy** [comment]
The nice thing is that it’s proprietary so at least nobody will use it in anything intended to be built on Hydra, and it does at least have a reason to exist. Some concerns, though:

* @TerrorJack Are you still interested in actively maintaining this package, in particular to keep up with security updates for all versions in a timely manner?

* Do we really need all the versions? Is anyone going to deploy JDK 8 with this?

* It seems like you need a paid licence to deploy it and that the free download only permits you to “use and evaluate the Product solely for its internal testing and development purposes”. They also want a sign‐up for that download – I am not sure we are adequately signposting that you cannot actually deploy this without paying them them “Request a Quote” money. I am also not sure if they are going to be happy with these deep links as opposed to, say, `requireFile`. This reminds me a bit of the Atlassian packages where we started shipping versions that required truly exorbitant licence fees to legally deploy. This is my main concern.

* The package name should perhaps be something like `azul-platform-prime` if we’re going to include this.

**Comment by TerrorJack** [comment]
Thanks for the thoughtful review @emilazy. As discussed in `#jvm:nixos.org`, the license concern is very legit and I intend to close this PR. Whoever wants to use azul's jvm in nixos might still be able to find the closed PR via search engine, and they are free to reuse the derivation here in their own derivation, and be responsible for updating jvm themselves :)

**Pull request closed without merging**